### PR TITLE
manual control stick input to velocity setpoint

### PR DESCRIPTION
**Pull request by Stifael**
- position control stick input to setpoint 
- reset clean up

Already flight tested several hours.  ~~It requires check for rebase errors~~

### CODE REVIEW

File: src/modules/mc_pos_control/mc_pos_control_params.c [hunk 1]
```diff
@@ -469,7 +481,31 @@ PARAM_DEFINE_FLOAT(MPC_ACC_UP_MAX, 5.0f);
  * @decimal 2
  * @group Multicopter Position Control
  */
-PARAM_DEFINE_FLOAT(MPC_ACC_DOWN_MAX, 5.0f);
+PARAM_DEFINE_FLOAT(MPC_ACC_DOWN_MAX, 2.0f);
```
Comment (MaEtUgR) [inline src/modules/mc_pos_control/mc_pos_control_params.c]: didn't you mention this was too low and that's why we increased it. the defaults should be resonable for upstream...

File: src/modules/mc_pos_control/mc_pos_control_params.c [hunk 2]
```diff
@@ -469,7 +481,31 @@ PARAM_DEFINE_FLOAT(MPC_ACC_UP_MAX, 5.0f);
...
+ * Maximum jerk in manual controlled mode for braking to zero
+ *
+ * @unit m/s/s/s
+ * @min 2.0
+ * @max 15.0
+ * @increment 1
+ * @decimal 2
+ * @group Multicopter Position Control
+ */
+PARAM_DEFINE_FLOAT(MPC_JERK_MAX, 5.0f);
```
Comment (MaEtUgR) [inline src/modules/mc_pos_control/mc_pos_control_params.c]: do you want toenable all smoothing features by default?
there are two sides to it:
- + the normal user who has a well supported airframe that just works probably wants smoothing
- + the developer can see what values make sense
- - the hobby user or developer that wants to tune the controller (which happens a lot I guess) has to study all these parameters to diable smoothing first. I myself currently don't remember all of them.

A nice solution would be one parameter that disables it all (e.g. set jerk to 0 or what not) or at least a guide that tells you exactly which steps to take...

Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_params.c]: yep, a guide is needed. I will add info about mc_pos_control parameters in general

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 3]
```diff
@@ -1320,12 +1320,18 @@ MulticopterPositionControl::control_manual(float dt)
...
-	set_manual_acceleration_xy(stick_xy, dt);
-	float stick_z = man_vel_sp(2);
-	float max_acc_z;
-	set_manual_acceleration_z(max_acc_z, stick_z, dt);
+	/* default for acceleration */
+	_acceleration_state_dependent_xy = _acceleration_hor_max.get();
+	_acceleration_state_dependent_z = _acceleration_z_max_up.get();
+	float max_acc_z = _acceleration_state_dependent_z;
+
+	/* adjust acceleration based on stick input only if jerk max < jerk min */
```
Comment (MaEtUgR) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: shouldn't it be normally min < max and if it's opposite you don't do anything?

**Comment by MaEtUgR** [comment]
@Stifael I rebased (the small conflict) on master and cherry picked my throttle reset fix from yesterday.
I thought maybe we merge this after the release...

**Comment by LorenzMeier** [comment]
@Stifael Please rebase on master.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 4]
```diff
@@ -177,10 +222,10 @@ class MulticopterPositionControl : public control::SuperBlock
 		param_t mc_att_yaw_p;
 		param_t hold_max_xy;
 		param_t hold_max_z;
-		param_t acc_up_max;
-		param_t acc_down_max;
 		param_t alt_mode;
 		param_t opt_recover;
+		param_t rc_flt_smp_rate;
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Could you please move these to the (new) blockparams?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 5]
```diff
@@ -435,6 +471,11 @@ MulticopterPositionControl::MulticopterPositionControl() :
 
 	_ref_pos = {};
 
+	/* set trigger time for manual direction change detection */
+	_manual_direction_change_hysteresis.set_hysteresis_time_from(false, DIRECTION_CHANGE_TRIGGER_TIME_US);
+
+	memset(&_ref_pos, 0, sizeof(_ref_pos));
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: This should be removed - its set to all-zero 3 lines above already.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 6]
```diff
@@ -449,6 +490,7 @@ MulticopterPositionControl::MulticopterPositionControl() :
 	_vel_sp_prev.zero();
 	_vel_err_d.zero();
 	_curr_pos_sp.zero();
+	_stick_input_xy_prev = matrix::Vector2f(0.0f, 0.0f);
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: use the zero initialiser: `_stick_input_xy_prev{}`. The code above constructs a zero-object matrix::Vector, then copies its contents into the zero-initialised _stick_input_xy_prev. Math example:

stick_input = 0;
zero_obj = 0;

stick_input = zero_obj;

On the other hand doing _stick_input_xy_prev does what you expect:

stick_input = 0;

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 7]
```diff
@@ -487,10 +529,10 @@ MulticopterPositionControl::MulticopterPositionControl() :
 
 	_params_handles.hold_max_xy = param_find("MPC_HOLD_MAX_XY");
 	_params_handles.hold_max_z = param_find("MPC_HOLD_MAX_Z");
-	_params_handles.acc_up_max = param_find("MPC_ACC_UP_MAX");
-	_params_handles.acc_down_max = param_find("MPC_ACC_DOWN_MAX");
 	_params_handles.alt_mode = param_find("MPC_ALT_MODE");
 	_params_handles.opt_recover = param_find("VT_OPT_RECOV_EN");
+	_params_handles.rc_flt_cutoff = param_find("RC_FLT_CUTOFF");
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Already commented above: Should use the new blockparam interface.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 8]
```diff
@@ -593,10 +635,16 @@ MulticopterPositionControl::parameters_update(bool force)
 		_params.hold_max_xy = math::max(0.f, v);
 		param_get(_params_handles.hold_max_z, &v);
 		_params.hold_max_z = math::max(0.f, v);
-		param_get(_params_handles.acc_up_max, &v);
-		_params.acc_up_max = v;
-		param_get(_params_handles.acc_down_max, &v);
-		_params.acc_down_max = v;
+		param_get(_params_handles.rc_flt_smp_rate, &(_params.rc_flt_smp_rate));
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: I see why you didn't use block param here. How about we move the constraints into the main code where the params are being used?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 9]
```diff
@@ -593,10 +635,16 @@ MulticopterPositionControl::parameters_update(bool force)
...
+		_params.rc_flt_smp_rate = math::max(1.0f, _params.rc_flt_smp_rate);
+		/* since we use filter to detect manual direction change, take half the cutoff of the stick filtering */
+		param_get(_params_handles.rc_flt_cutoff * 0.7f, &(_params.rc_flt_cutoff));
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: This is not working. By multiplying the param handle its value gets trashed.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 10]
```diff
@@ -593,10 +635,16 @@ MulticopterPositionControl::parameters_update(bool force)
...
+		/* make sure the filter is in its stable region -> fc < fs/2 */
+		_params.rc_flt_cutoff = math::constrain(_params.rc_flt_cutoff, 0.1f, (_params.rc_flt_smp_rate / 2.0f) - 1.f);
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: This led to a correct value. But the system never ever used rc_flt_cutoff.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 11]
```diff
@@ -1038,21 +1395,10 @@ MulticopterPositionControl::control_manual(float dt)
...
-		_att_sp.roll_body = 0.0f;
-		_att_sp.pitch_body = 0.0f;
-		_att_sp.yaw_body = _yaw;
-		_att_sp.thrust = 0.0f;
-
-		_att_sp.timestamp = hrt_absolute_time();
+		/* don't run controller when landed
+		 * NOTE:
+		 * This only works in manual since once we give throttle input, the
+		 * landdetector will exit the landing state */
```
Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Does the attitude controller command zero rates when landed?

Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: need to check. but the position controller sends zero thrust if landed

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 10 above)
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: yes, need to set it as well, but probably it is best to do it within the sensor module. 
anyway, your comments are noted and written down and will add it to a different pr

**Comment by r0gelion** [comment]
@LorenzMeier is this ready for flight testing?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 11 above)
Comment (MaEtUgR) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: I don't think it does. Only XY position control get disabled because they produced the biggest problem with tip overs and too high thrust when trying to land because the vehicle does not move anymore.

**Comment by MaEtUgR** [comment]
I think @Stifael is currently testing the newest changes. I think he should decide when it's open for the general acceptance test.

**Comment by Stifael** [comment]
@r0gelion , you can go ahead and test it.
However, I think I need to explain first what the changes incorporate, otherwise it might be difficult to test (the devguide need to be updated for that).

The changes effect position control handling:
the important parameters that effect the manual position control performance are the following:
MPC_ACC_HOR_MAX (default = 10): defines how fast the vehicle will stop
MPC_ACC_HOR (default = 5): defines how fast the vehicle accelerates
MPC_DEC_HOR_SLOW (default = 5): defines how fast the vehicle decelerates
MPC_JERK_MAX(default  =0): defines how smooth the vehicle should transition from fast forward flight to stop
MPC_JERK_MIN(default=1): defines how smooth the vehicle should transition from forward flight to stop
* If MPC_JERK_MIN > MPC_JERK_MAX, then there is an immediate transition (MPC_ACC_HOR goes directly to MPC_ACC_HOR_MAX when sticks are at zero position like it is in the current master)
* if MC_JERK_MAX > MPC_JERK_MIN, the transition from MPC_ACC_HOR to MPC_ACC_HOR_MAX occurs smooth with a value in between  MC_JERK_MAX and  MPC_JERK_MIN dependent on the velocity


1. Start with default values
If you use the default values and fly around in position control, the difference to the current master is the following:
* there is no delay if you do direction change or put sticks to zero 

2. decrease MPC_DEC_HOR_SLOW to 1
The vehicle will now accelerate much slower when you move the stick. The acceleration is linearly adjusted to the stick input (at max stick, acceleration is MPC_ACC_HOR, at very low stick input, the acceleration is MPC_DEC_HOR). However, note that the vehicle will still stop abruptly when moving the stick to zero. You can test this by providing a low stick input which results in a very slow forward flight and then apply zero stick input, which results in an abrupt stop.

3. increase MPC_JERK_MAX to some value above MPC_JERK_MIN (for instance 15)
The vehicle will behave the same as in 2, but the stop is not abrupt anymore.

It is to note that these values above are tuning values on top of the normal PID turning values, and depend a lot on the vehicle.

**Comment by Avysuarez** [comment]
Flights with Pixhawk 1 (V3); Good flights no issues, The result of the tests is the same as the indications.

1. Default values.
https://review.px4.io/plot_app?log=d43660e5-5fbb-49b6-89a7-ea700d229a6a

2. Decrease MPC_DEC_HOR_SLOW to 1.
https://review.px4.io/plot_app?log=9f7b43dd-0859-4369-8a22-d225803016de

3. Increase MPC_JERK_MAX .
https://review.px4.io/plot_app?log=d28cb85f-138f-4f77-b62f-241cda024a80

**Comment by bresch** [comment]
@Stifael Nice work, but I have a few questions and remarks to do:

1. Since the file `mc_pos_control_main.cpp` is already quite large, what would you think about implementing your "manual control stick input to velocity setpoint" in a class used by the MPC? This would limit the expansion of the ~universe~ position controller class.

2. Your algorithm limits the jirk when the velocity is in the interval `]0,XY_VEL_MAX[`. However, as soon as the velocity setpoint reaches the saturation (**or zero!**), the setpoint is clamped and a Dirac jerk is generated. This means that if your velocity controller is well tuned, the multicopter will have an abrupt change of attitude when you reach the max speed and when you reach zero speed. The latter is particularly concerning.
xyspeed_sp_zero

3. I don't fully understand why you use two different decelerations for decelerating and breaking. The jerk limitation will smooth out the spikes you would see while decelerating. Furthermore, why should a change of speed be slower than a breaking action?

4. During `direction_change`, you adjust the acceleration with the stick input. Why? Isn't the stick input supposed to define the velocity target?

I would like to understand mode in depth how you designed this mapping behavior. Any references to an existing technique? Maybe could you share a document explaining the design requirements and the choices you made?

Many thanks!

**Comment by MaEtUgR** [comment]
@bresch 1. is under construction, I'm currently already working on porting this functionality into a flight task called FlightTaskManual which is actually a separate class (see https://github.com/PX4/Firmware/pull/7381/commits/39f15c14683de93247d65be9b61c48dcb90faef8) for first attempts which are highly WIP and not usable yet.

**Comment by Stifael** [comment]
@bresch, i will try to go through the changes one by one and hopefully that will make it easier to understand the changes. In fact, the changes are not too complicated and are implemented directly based on user feedback and requirements.

Let's start from the state of having just the position/velocity controller tuned **without** any additional acceleration (aka slewrate) and jerk limits active:
In the current tap and in this branch the stick inputs range from 0 to 1 (except thrust that ranges from -1 to 1, but that is another topic). Those stick inputs are mapped to velocity setpoints with 1 = maximum horizontal velocity and 0 = 0 velocity, where the map does not occur linearly but exponentially (see [here](https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L927-L940)). Having just this stick to velocity map is already enough if you just want to fly around and not planing on flying smooth trajectories for video filming for instance. Direct stick inputs to velocity setpoint result in step inputs. For instance, if the stick inputs map to 1m/s, and then the sticks are moved such that it maps to 5m/s, this will be seen by the controller as an almost step input from 1 to 5. In order to deal with it setpoint slewrate was introduced ([here](https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L1255-L1278)). The slewrate defines how fast a setpoint can change. Going back to the 1 to 5 m/s example, the slewrate prevents the velocity setpoint jumping immediately from 1m/s to 5m/s but instead linearly goes from 1m/s to 5m/s over time. The smaller the acceleration limit is chosen, the longer it takes for the sepoint to go from 1m/s to 5m/s. The variable that controls this change is MPC_ACC_HOR_MAX and can be considered as a tuning parameter. However, there are several **problems with slewrate** if it is just blindly applied to change the rate of velocity setpoint:

**velocity setpoint slewrate is feedforward-like**
This means it does not take the current velocity state into account. Problems that arise with such a feedforward setpoint generation is that it introduces a delay in setpoint which is primarily severe if the user wants to do a direction change or a brake that only occurs once the velocity setpoint reached the current velocity state (an example is given [here](https://github.com/PX4/Firmware/issues/6681#issue-210436062)). For instance, if the user demands 10m/s from hover, the velocity setpoint will linearly increase over time from 0 to 10m/s, where the actual velocity of the vehicle is possibly much lower than the velocity setpoint (especially if the vehicle actual dynamics is much slower than the demand in velocity change). lets assume that the user demands a brake (zero stick input) as soon as the velocity setpoint reached 10m/s. If now the actual vehicle velocity is only 5m/s, the vehicle will continue to accelerate in the same direction until the velocity setpoint decreased from 10m/s below the actual velocity. It gets even worse if we introduce an other variable that defines how fast the velocity setpoint can change during deceleration, which is usually set lower than acceleration (the reason for having it lower is again user feedback: if you fly forward, the user does not want to see the vehicle jerking hard when decreasing the stick inputs slightly). 

In order to deal with the delay one solution is to reset the slewrate when the user demands to brake or direction change. However, the user intention cannot be derived from the velocity septoint due to the delay introduced by the slewrate. On the other hand, the intention can be derived from the stick inputs ([here](https://github.com/PX4/Firmware/pull/7242/files#diff-fc77c3ef569029d45764664c75d4b0c1R1114)).

At this stage, we now have a slewrate that is defined by one or two variables (acceleration and/or deceleration). The problem with that is, however, that the allowed change in velocity setpoint is independent of the user demand. For instance, if the user just gives a small stick input, he expects that the vehicle accelerates slowly, whereas at large stick input the user expects that vehicle accelerates fast. To satisfy the user expectation, this PR couples linearly the acceleration limit used by the slewrate to the stick inputs. If now the user demands 0.5m/s from hover, the velocity setpoint changes very slowly until it reaches 0.5m/s, which results in a very slow actual acceleration. In comparison, if the user demands 10m/s from hover, the velocity setpoint changes fast and therefore the vehicle accelerates fast as well.

The above mapping from stick inputs to velocity setpoint and velocity setpoint change applies if the stick inputs are **non-zero**. However, if the stick inputs are at zero and therefore the user wants to brake or hold position, the acceleration limit used by the slewrate should be as large as possible. **Why**: In contrast to fast forward flight, during loiter the controller uses position controller cascaded to the velocity controller. Since the slewrate introduces a delay, the output of the position controller (= input to velocity controller) would be delayed as well, which results in bad hover flight. As a consequence, during hover the acceleration limit should be set quite large or not using acceleration limit at all. In this PR, the acceleration limit during hover is set to the max acceleration MPC_ACC_HOR_MAX ([here](https://github.com/PX4/Firmware/pull/7242/files#diff-ffdc6fb3dc66ee0fd3359190869b8712R412)), which should be set larger than the max acceleration during forward flight ([here](https://github.com/PX4/Firmware/pull/7242/files#diff-ffdc6fb3dc66ee0fd3359190869b8712R424)). 

Since MPC_ACC_HOR_MAX > (acceleration limit from stick input), we still need to take care of the transition from forward flight to brake. For instance, assume the user demands a forward flight of 1m/s. in addition, also assume that this maps to 1m/s*s acceleration. If now the user demands a brake, the acceleration limit jumps immediately from 1m/ss to MPC_ACC_HOR_MAX. This results in a very abrupt, fast stop, which is not always desired. In fact, the user expects a very smooth brake at low speed and fast brake at large speed. This is where jerk limit comes in. The **jerk** limit is only active and used for **brake** ([here[(https://github.com/PX4/Firmware/pull/7242/files#diff-fc77c3ef569029d45764664c75d4b0c1R1189)). When braking, the value of the jerk limit depends linearly on the acceleration limit. For instance, if the user gives full stick input and therefore the acceleration limit is at maximum horizontal acceleration for manual ([here](https://github.com/PX4/Firmware/pull/7242/files#diff-ffdc6fb3dc66ee0fd3359190869b8712R424)) and then the user demands a brake, the jerk will be at maximum jerk and therefore the acceleration linearly increases over time from MPC_ACC_HOR to the hover acceleration MPC_ACC_HOR_MAX with rate MPC_JERK_MAX. In contrast, if the user demands low speed and therefore the acceleration limit is low, a demanded brake would be slower because the mapped low acceleration limit will approach MPC_ACC_HOR_MAX slowly over time with a jerk rate closer to MPC_JERK_MIN. 

To sum up what this PR does:
- derive intention from stick input
- reset slewrate for brake and direction change to overcome delay
- adjust acceleration depending on stick input
- smooth brake by adding jerk wihich depends on acceleration limit when brake is initiated

I hope this answers some of your questions.

**Comment by Stifael** [comment]
I close this PR.
It is simpler to squash merge and then do a rebase

**Comment by bresch** [comment]
@Stifael Thanks for clarification, I'll try to have a deeper look on the code later.

**Comment by vmishra9** [comment]
Are the JERK parameter used in auto mode too? I have a big quadcopter which starts oscillating when it is close to reaching a way point. This is happening only after I upgraded to v.1.7.2. Before this I had v.1.5.4 and I did not have any problems. 
I retuned the gains after I upgraded to v.1.7.2 and the impulse response of the quadcopter was acceptable. The problem appears only when it starts breaking in the auto mode.

**Comment by Stifael** [comment]
@vmishra9 no, in auto mode there are no smoothing applied. 
Do you have log?

**Comment by vmishra9** [comment]
@Stifael Here is the log of the flight:

https://logs.px4.io/plot_app?log=62935c55-ff5a-4345-afd6-1e38b801bc57

**Comment by vmishra9** [comment]
@Stifael I did another trial after retuning the ROLLRATE_P and PITCHRATE_P and decreasing it by a large amount. The flight was good initially but again after the first way point the quad started oscillating. Here is the log

https://logs.px4.io/plot_app?log=84a29046-efc0-4686-82c8-5ffc644b228a
